### PR TITLE
fix(eslint-plugin): [unified-signatures] deduplicate types in report

### LINKED ISSUE #12504: Bug: [unified-signatures] Error message contains the same type twice when it appears in more than one signature
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.
- [x] If I used AI assistance to prepare this issue, I have reviewed and followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/).

### Playground Link

https://typescript-eslint.io/play/#ts=6.0.2&fileType=.ts&code=GYVwdgxgLglg9mABMAFAQwFyLCAtgIwFMAnRAH0QGcpiYwBzASiwDc4YATAbgChRJYCZOizVaDconxw4AG0JowzRG05cgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJygM0sQBNaySgHMmAQ3yxoKdGADa4bDkTRoAe2iQANIqVYl2SCKYbEAEW5cViFvEIA5MQFt%2BABTHRnifCtQZ8cIi62AC%2BwQC6imEhQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
function f(a: number | string): void;
function f(a: string | boolean): void;
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/unified-signatures": [
      "error",
      {
        "ignoreDifferentlyNamedParameters": true
      }
    ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Should error with:

> This overload and the one on line 1 can be combined into one signature taking `number | string | boolean`

### Actual Result

Errors with:

> This overload and the one on line 1 can be combined into one signature taking `number | string | string | boolean`

(notice that `string` appears twice)

### Additional Info

Note that it might not be that straightforward to fix since we'll need to somehow parse the "final" type and remove duplicates from it.
Would be nice if we could have some TS API that takes the whole thing, "`eval`s" it, and calls `.toString()`

Co-authored-by: @JoshuaKGoldberg when reviewing #12143 -> #12493 together

### PR DESCRIPTION
**Pull request by nayounsang**
## PR Checklist

- [x] Addresses an existing open issue: fixes #12504
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken
- [x] If this PR used AI assistance, I followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/)

#### Unify
- Generic type parameters are declared in a separate scope for each overload. Therefore, even if both overloads declare a type parameter named `T`, comparing scope bindings alone treats them as different declarations.
- To handle this, the `Unify` result now stores both `typeParameters0` from the first signature and `typeParameters1` from the second signature.
- When generating the message, each type reference is checked against the type parameters of its corresponding signature.

#### Message
- Previously, the message template assumed exactly two distinct types and constructed the union as `type1 | type2`.
- With this change, each type can already be a union, and flattening and deduplication can result in anywhere from one to multiple union members.
- the rule now constructs the final union string as a single `types` value, and the message renders that value directly. This does not change the user-facing message.

### Implement
1. `getUnifiedTypeText`: Converts the two parameter types into union members for the diagnostic message.
2. `getUnionMembers`: Recursively flattens only `TSUnionType` nodes.
3. `typeNodesAreEqual`: Determines whether the flattened members are duplicates.
   - For regular types, compares their AST structure.
   - For type references, also compares their scope bindings.
   - For generic type parameters declared per signature, uses the provided declarations to treat type parameters at the same index as corresponding to each other.
4. `getUnifiedTypeText`: Removes duplicates and constructs the final union type string.
5. `getUnionMemberText`: Retrieves the source text for each member to be used in the final message, adding parentheses when necessary.

### Test
- test modifications due to interface changes

### Additional Info
- I left a question in the code comments.

### CODE REVIEW

File: packages/eslint-plugin/tests/rules/unified-signatures.test.ts [hunk 1]
```diff
@@ -1336,5 +1312,221 @@ function f(this: string | number, a: boolean): void {}
...
+          messageId: 'singleParameterDifference',
+        },
+      ],
+    },
+    {
+      // An alias and the type it denotes are distinct syntax and remain distinct.
+      code: `
+type Name = string;
+function f(x: Name): void;
+function f(x: string | Name): void;
```
Comment (nayounsang) [inline packages/eslint-plugin/tests/rules/unified-signatures.test.ts:1400]: We need to make a decision regarding this policy.
Semantically, they are the same type, but the type names differ.
However, supporting this would likely entail several implementations.((It is just my prediction.)
If the types are semantically equivalent, I suggest selecting the type that appears first (or last). Alternatively, simply not addressing the issue.

File: packages/eslint-plugin/src/rules/unified-signatures.ts [hunk 2]
```diff
@@ -295,10 +304,206 @@ export default createRule<Options, MessageIds>({
...
+    }
+
+    function getUnifiedTypeText(
+      type0: TSESTree.TypeNode | undefined,
+      type1: TSESTree.TypeNode | undefined,
+      typeParameters0?: TSESTree.TSTypeParameterDeclaration,
+      typeParameters1?: TSESTree.TSTypeParameterDeclaration,
+    ): string {
+      if (type0 == null || type1 == null) {
+        return '';
```
Comment (nayounsang) [inline packages/eslint-plugin/src/rules/unified-signatures.ts]: This return is strange. 
```ts
     function f(value): void;
     function f(value: string): void;
```
I consider this situation, but I’m wondering if there might be a better alternative.
How should We handle something like this?

I might consider a no-op in this case.

File: packages/eslint-plugin/tests/rules/unified-signatures.test.ts
Comment (StyleShit) [inline packages/eslint-plugin/tests/rules/unified-signatures.test.ts:1]: [Praise]

Great test cases!

File: packages/eslint-plugin/src/rules/unified-signatures.ts [hunk 3]
```diff
@@ -130,24 +141,20 @@ export default createRule<Options, MessageIds>({
...
             const lineOfOtherOverload = only2 ? undefined : p0.loc.start.line;
 
-            const typeAnnotation0 = isTSParameterProperty(p0)
-              ? p0.parameter.typeAnnotation
-              : p0.typeAnnotation;
-            const typeAnnotation1 = isTSParameterProperty(p1)
-              ? p1.parameter.typeAnnotation
-              : p1.typeAnnotation;
+            const type0 = getParameterType(p0);
+            const type1 = getParameterType(p1);
```
Comment (StyleShit) [inline packages/eslint-plugin/src/rules/unified-signatures.ts]: [Nitpick, non-blocking]

I think this is nicer/easier to follow?

```suggestion
            const typeAnnotation0 = getParameterTypeAnnotation(p0);
            const typeAnnotation1 = getParameterTypeAnnotation(p1);
```

File: packages/eslint-plugin/src/rules/unified-signatures.ts (hunk 2 above)
Comment (StyleShit) [reply packages/eslint-plugin/src/rules/unified-signatures.ts]: Hmm

The first signature is basically redundant, so maybe let's treat it as such?
In other words - tell the user to replace both of these signatures with `value: string`, which is what I think I would've done manually in such cases

Unless other team members have a different opinion

File: packages/eslint-plugin/src/rules/unified-signatures.ts [hunk 4]
```diff
@@ -295,10 +304,212 @@ export default createRule<Options, MessageIds>({
...
+      const text = getTextWithParentheses(context.sourceCode, type);
+      const needsParentheses =
+        type.type === AST_NODE_TYPES.TSConditionalType ||
+        type.type === AST_NODE_TYPES.TSConstructorType ||
+        type.type === AST_NODE_TYPES.TSFunctionType;
+
+      return needsParentheses && !isParenthesized(type, context.sourceCode)
+        ? `(${text})`
+        : text;
+    }
```
Comment (StyleShit) [inline packages/eslint-plugin/src/rules/unified-signatures.ts:347]: [Bug]

This will catch parenthesized expressions that contain comments inside them:

```TS
declare function fn(a: number): void;
declare function fn(a: (/* before */ string /* after */)): void;

// These overloads can be combined into one signature taking `number | (/* before */ string /* after */)`.
```

[Playground](https://deploy-preview-12656--typescript-eslint.netlify.app/play/#ts=6.0.2&fileType=.tsx&code=CYUwxgNghgTiAEAzArgOzAFwJYHtVNQAooAueVZAWwCMQYBKMgNxy2AG4AoUSWBFdNjwFiZQgHoAVPFqIcceJPHwAzhhhZUAc3hT4URBjqLx9RvBZt2QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYkACAXAngA4oDG0AlobgLQrzkB2uA9LA%2BQGbmIAm1y5AOYMAhrljQU6KImjQA9tEgBfEMqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

File: packages/eslint-plugin/src/rules/unified-signatures.ts [hunk 5]
```diff
@@ -295,10 +304,212 @@ export default createRule<Options, MessageIds>({
...
+
+    function typeNodesAreEqual(
```
Comment (StyleShit) [inline packages/eslint-plugin/src/rules/unified-signatures.ts]: [Simplification]

Can we go with something like `typesAreEqual` that we have in the same file?

https://github.com/typescript-eslint/typescript-eslint/blob/c269b34c0c29c279812bc7f415af451447d4ff27/packages/eslint-plugin/src/rules/unified-signatures.ts#L490-L501

We try not to recreate/duplicate the compiler logic here

**Review by StyleShit — CHANGES_REQUESTED** [review]
That's a great start!

Added a couple of comments, most notably around missing test cases and simplification

Thanks for working on this ✨

File: packages/eslint-plugin/src/rules/unified-signatures.ts (hunk 5 above)
Comment (nayounsang) [reply packages/eslint-plugin/src/rules/unified-signatures.ts]: These two helpers serve different purposes. Therefore, it is difficult to reuse.

`typesAreEqual` compares whether two TSTypeAnnotations use the same type expression based on their source text.

In contrast, the union-member comparison addressed by this issue needs to account for generic type-parameter scopes and AST structure, so **comparing source text alone is insufficient.** Therefore, `typeNodesAreEqual` is necessary to preserve this comparison logic.

If we were to replace the existing usages of `typesAreEqual` with `typeNodesAreEqual`, it seems like it would become significantly heavier.

File: packages/eslint-plugin/src/rules/unified-signatures.ts (hunk 2 above)
Comment (nayounsang) [reply packages/eslint-plugin/src/rules/unified-signatures.ts]: [9f14ae4](https://github.com/typescript-eslint/typescript-eslint/pull/12656/commits/9f14ae4ebeac217d55b5bc026a22bac8d28eab1f)

To-be is `string`.  Keeping only one type. It is impossible for both to be null.

File: packages/eslint-plugin/src/rules/unified-signatures.ts (hunk 3 above)
Comment (nayounsang) [reply packages/eslint-plugin/src/rules/unified-signatures.ts]: I committed. [8816cfa](https://github.com/typescript-eslint/typescript-eslint/pull/12656/commits/8816cfaa785dc9b878e781223be791c93febfdb9)

File: packages/eslint-plugin/src/rules/unified-signatures.ts (hunk 4 above)
Comment (nayounsang) [reply packages/eslint-plugin/src/rules/unified-signatures.ts:347]: I applied a fix to remove the comments. [6f1ec28](https://github.com/typescript-eslint/typescript-eslint/pull/12656/commits/6f1ec28a996d3cc887acebf3b04733a1dba3d84b)

File: packages/eslint-plugin/tests/rules/unified-signatures.test.ts (hunk 1 above)
Comment (nayounsang) [reply packages/eslint-plugin/tests/rules/unified-signatures.test.ts:1400]: Handling this implies that deep comparisons must also be performed for deeply nested types.

File: packages/eslint-plugin/src/rules/unified-signatures.ts (hunk 5 above)
Comment (StyleShit) [reply packages/eslint-plugin/src/rules/unified-signatures.ts]: I don't mean to use the exact same function, but to use something similar (even though I _think_ we can reuse the original one with a small refactor)

Anyway, this is what I mean:

```diff
diff --git a/packages/eslint-plugin/src/rules/unified-signatures.ts b/packages/eslint-plugin/src/rules/unified-signatures.ts
index 1bc26275b..e08ec25cb 100644
--- a/packages/eslint-plugin/src/rules/unified-signatures.ts
+++ b/packages/eslint-plugin/src/rules/unified-signatures.ts
@@ -338,35 +338,17 @@ export default createRule<Options, MessageIds>({
         return getUnionMemberText(type0);
       }
 
-      const members = [
-        ...getUnionMembers(type0).map(type => ({
-          type,
-          typeParameters: typeParameters0,
-        })),
-        ...getUnionMembers(type1).map(type => ({
-          type,
-          typeParameters: typeParameters1,
-        })),
-      ];
+      const members = [...getUnionMembers(type0), ...getUnionMembers(type1)];
       const uniqueMembers: typeof members = [];
 
       for (const member of members) {
-        if (
-          !uniqueMembers.some(other =>
-            typeNodesAreEqual(
-              other.type,
-              member.type,
-              other.typeParameters,
-              member.typeParameters,
-            ),
-          )
-        ) {
+        if (!uniqueMembers.some(other => typeNodesAreEqual(other, member))) {
           uniqueMembers.push(member);
         }
       }
 
       return uniqueMembers
-        .map(member => getUnionMemberText(member.type))
+        .map(member => getUnionMemberText(member))
         .join(' | ');
     }
 
@@ -391,9 +373,12 @@ export default createRule<Options, MessageIds>({
     function typeNodesAreEqual(
       a: TSESTree.TypeNode,
       b: TSESTree.TypeNode,
-      typeParameters0?: TSESTree.TSTypeParameterDeclaration,
-      typeParameters1?: TSESTree.TSTypeParameterDeclaration,
     ): boolean {
+      return (
+        a === b ||
+        context.sourceCode.getText(a) === context.sourceCode.getText(b)
+      );
+
       const typeParameterIndices0 = new Map(
         typeParameters0?.params.map((parameter, index) => [
           parameter.name.name,

```

All existing tests are passing for me locally, but it does fail with this new test case though (because `.getText` is not the same):

```TS
declare function fn(a: number & { __brand: true }): void;
declare function fn(a: string | (number & /* */  { __brand: true })): void;
```

IMO it's okay to make this trade-off (of a very edgy edge case) to avoid writing a huge block of logic, and because it only affects the message and not the behavior of the rule

File: packages/eslint-plugin/tests/rules/unified-signatures.test.ts (hunk 1 above)
Comment (StyleShit) [reply packages/eslint-plugin/tests/rules/unified-signatures.test.ts:1400]: Let's check only the names without following the underlying type, as it'll require the use of the type checker.

See the discussion in https://github.com/typescript-eslint/typescript-eslint/issues/12505

File: packages/eslint-plugin/src/rules/unified-signatures.ts (hunk 5 above)
Comment (nayounsang) [reply packages/eslint-plugin/src/rules/unified-signatures.ts]: > IMO it's okay to make this trade-off (of a very edgy edge case) to avoid writing a huge block of logic, and because it only affects the message and not the behavior of the rule

Yes, I understand the review! Then, `typesAreEqual` becomes reusable with a slight change. I comitted.

P.S. Since I cannot request a re-review, I am sending this ping instead. @StyleShit

File: packages/eslint-plugin/src/rules/unified-signatures.ts [hunk 6]
```diff
@@ -295,10 +292,69 @@ export default createRule<Options, MessageIds>({
       // See https://github.com/Microsoft/TypeScript/issues/5077
       return parametersHaveEqualSigils(a, b) &&
         a.type !== AST_NODE_TYPES.RestElement
-        ? { kind: 'single-parameter-difference', p0: a, p1: b }
+        ? {
+            kind: 'single-parameter-difference',
+            p0: a,
+            p1: b,
+          }
```
Comment (StyleShit) [inline packages/eslint-plugin/src/rules/unified-signatures.ts]: [Nipick]

Let's revert this back to a one-liner, just to have less unrelated diff

File: packages/eslint-plugin/src/rules/unified-signatures.ts [hunk 7]
```diff
@@ -295,10 +292,69 @@ export default createRule<Options, MessageIds>({
...
     }
 
+    function getParameterTypeAnnotation(
+      parameter: TSESTree.Parameter,
+    ): TSESTree.TypeNode | undefined {
+      return (
+        isTSParameterProperty(parameter)
+          ? parameter.parameter.typeAnnotation
+          : parameter.typeAnnotation
+      )?.typeAnnotation;
```
Comment (StyleShit) [inline packages/eslint-plugin/src/rules/unified-signatures.ts]: [Nitpick, Style]

I think this is easier to read (subjective, of course)?

```suggestion
      return isTSParameterProperty(parameter)
        ? parameter.parameter.typeAnnotation?.typeAnnotation
        : parameter.typeAnnotation?.typeAnnotation;
```

But feel free to ignore me, that's basically personal preference, and not sure other devs will agree with me

File: packages/eslint-plugin/src/rules/unified-signatures.ts [hunk 8]
```diff
@@ -295,10 +292,69 @@ export default createRule<Options, MessageIds>({
...
+      type1: TSESTree.TypeNode | undefined,
+    ): string {
+      if (type0 == null) {
+        return getUnionMemberText(
+          nullThrows(type1, 'Expected the other parameter type to be defined'),
+        );
+      }
+      if (type1 == null) {
+        return getUnionMemberText(type0);
+      }
```
Comment (StyleShit) [inline packages/eslint-plugin/src/rules/unified-signatures.ts]: These 2 can be unified into a single condition (also added a comment. feel free to reword):

```suggestion
      // When a signature's parameter has no type annotation
      if (type0 == null || type1 == null) {
        return getUnionMemberText(
          nullThrows(
            type0 ?? type1,
            'Expected a type annotation for one of the parameters, but both were undefined',
          ),
        );
      }
```

File: packages/eslint-plugin/src/rules/unified-signatures.ts [hunk 9]
```diff
@@ -295,10 +292,69 @@ export default createRule<Options, MessageIds>({
...
+
+      const members = [...getUnionMembers(type0), ...getUnionMembers(type1)];
+      const uniqueMembers: typeof members = [];
```
Comment (StyleShit) [inline packages/eslint-plugin/src/rules/unified-signatures.ts]: [Nitpick]

```suggestion
      const uniqueMembers: TSESTree.TypeNode[] = [];
```

File: packages/eslint-plugin/src/rules/unified-signatures.ts [hunk 10]
```diff
@@ -295,10 +292,69 @@ export default createRule<Options, MessageIds>({
...
+    }
+
+    function getUnionMemberText(type: TSESTree.TypeNode): string {
+      const text = context.sourceCode.getText(type);
+      const needsParentheses =
+        type.type === AST_NODE_TYPES.TSConditionalType ||
+        type.type === AST_NODE_TYPES.TSConstructorType ||
+        type.type === AST_NODE_TYPES.TSFunctionType;
+
+      return needsParentheses || isParenthesized(type, context.sourceCode)
```
Comment (StyleShit) [inline packages/eslint-plugin/src/rules/unified-signatures.ts]: I think something is off with the wording here?

- If a node `needsParentheses`, add parentheses --> fine
- If a node `isParenthesized`, add parentheses --> ??

File: packages/eslint-plugin/src/rules/unified-signatures.ts [hunk 11]
```diff
@@ -488,15 +544,20 @@ export default createRule<Options, MessageIds>({
...
     function typesAreEqual(
-      a: TSESTree.TSTypeAnnotation | undefined,
-      b: TSESTree.TSTypeAnnotation | undefined,
+      a: TSESTree.TSTypeAnnotation | TSESTree.TypeNode | undefined,
+      b: TSESTree.TSTypeAnnotation | TSESTree.TypeNode | undefined,
     ): boolean {
+      const typeA =
+        a?.type === AST_NODE_TYPES.TSTypeAnnotation ? a.typeAnnotation : a;
+      const typeB =
+        b?.type === AST_NODE_TYPES.TSTypeAnnotation ? b.typeAnnotation : b;
```
Comment (StyleShit) [inline packages/eslint-plugin/src/rules/unified-signatures.ts]: All tests pass if you remove these conditions. Are they necessary? Or maybe there are missing tests?

Also, I think it makes more sense to pass only `TypeNode`s to it and remove the `TSTypeAnnotation`, if possible

File: packages/eslint-plugin/tests/rules/unified-signatures.test.ts [hunk 12]
```diff
@@ -1336,5 +1312,377 @@ function f(this: string | number, a: boolean): void {}
         },
       ],
     },
+    {
+      // https://github.com/typescript-eslint/typescript-eslint/issues/12504
```
Comment (StyleShit) [inline packages/eslint-plugin/tests/rules/unified-signatures.test.ts]:
```suggestion
```

I don't think we need this

File: packages/eslint-plugin/tests/rules/unified-signatures.test.ts [hunk 13]
```diff
@@ -1336,5 +1312,377 @@ function f(this: string | number, a: boolean): void {}
...
+          },
+          endColumn: 57,
+          endLine: 3,
+          line: 3,
+          messageId: 'singleParameterDifference',
+        },
+      ],
+    },
+    {
+      // To avoid complex type comparisons when resolving #12504, allows duplicate union members when comments make their source text differ.
```
Comment (StyleShit) [inline packages/eslint-plugin/tests/rules/unified-signatures.test.ts]:
```suggestion
      // To avoid complex type comparisons when resolving #12504, allow duplicate union members when their source text differs.
```

It can also differ due to spacing and such

File: packages/eslint-plugin/tests/rules/unified-signatures.test.ts [hunk 14]
```diff
@@ -1336,5 +1312,377 @@ function f(this: string | number, a: boolean): void {}
...
+      code: noFormat`
+function f(x: string & { brand: true }): void;
+function f(x: number | string & /* brand */ { brand: true }): void;
```
Comment (StyleShit) [inline packages/eslint-plugin/tests/rules/unified-signatures.test.ts:1501]: [Question]

Hmm, should we `.replace('\n', '')` all the members' text to avoid weird cases like this:

```
These overloads can be combined into one signature taking `string | number & {
	brand: true
}`.
```

**Review by StyleShit — CHANGES_REQUESTED** [review]
This is looking very good! Just some final touches ⚡️

File: packages/eslint-plugin/tests/rules/unified-signatures.test.ts (hunk 13 above)
Comment (nayounsang) [reply packages/eslint-plugin/tests/rules/unified-signatures.test.ts]: That's right. However, since that falls under the formatter's role, I won't add a test case for it.

File: packages/eslint-plugin/src/rules/unified-signatures.ts (hunk 11 above)
Comment (nayounsang) [reply packages/eslint-plugin/src/rules/unified-signatures.ts]:
```ts
function f(value: number): string;
function f(value: string):    string;

interface I {
  f(x: number): void;
  f(x:    number, extra?: string): void;
}
```
This conditional statement properly handles such formatting differences. Therefore, the conditional statement itself is necessary.
I will extract the `TypeNode` at the existing call. The behavior doesn't change, since it's an interface change.

File: packages/eslint-plugin/src/rules/unified-signatures.ts (hunk 10 above)
Comment (nayounsang) [reply packages/eslint-plugin/src/rules/unified-signatures.ts]: At the time, I intended to preserve the parentheses for types containing comments, but since I was unsure about that myself, I think it’s fine to remove them.

File: packages/eslint-plugin/tests/rules/unified-signatures.test.ts (hunk 14 above)
Comment (nayounsang) [reply packages/eslint-plugin/tests/rules/unified-signatures.test.ts:1501]: I think it's nice to respect the existing codebase and there might be cases where putting it all on a single line actually makes it harder to read.

File: packages/eslint-plugin/tests/rules/unified-signatures.test.ts [hunk 15]
```diff
@@ -1476,27 +1475,7 @@ function f(x: number | /* second */ string): void;
...
-              'These overloads can be combined into one signature',
-            types: 'number | (string)',
-          },
-          endColumn: 57,
-          endLine: 3,
-          line: 3,
-          messageId: 'singleParameterDifference',
-        },
-      ],
-    },
```
Comment (StyleShit) [inline packages/eslint-plugin/tests/rules/unified-signatures.test.ts]: This is a good test, why did you remove it?
It should've just been fixed to `types: 'number | string'`, no?

**Review by StyleShit — CHANGES_REQUESTED** [review]
Everything is much cleaner!
Just one comment about the removed test

File: packages/eslint-plugin/tests/rules/unified-signatures.test.ts (hunk 15 above)
Comment (nayounsang) [reply packages/eslint-plugin/tests/rules/unified-signatures.test.ts]: yes. I deleted it because I thought it was a duplicate test case if () wasn't included, but I will revert that change.

**Review by StyleShit — APPROVED** [review]
Thanks!

**Review by JoshuaKGoldberg — APPROVED** [review]
💯
